server: TestTelemetrySQLStatsIndependence failed

[(server).TestTelemetrySQLStatsIndependence failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2194255&tab=buildLog) on [master@22e1183487917118e189774ed938145d6c45d9a3](https://github.com/cockroachdb/cockroach/commits/22e1183487917118e189774ed938145d6c45d9a3):

```

Goroutine 865 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:960 +0x651
  testing.runTests.func1()
      /usr/local/go/src/testing/testing.go:1202 +0xa6
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:909 +0x199
  testing.runTests()
      /usr/local/go/src/testing/testing.go:1200 +0x521
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:1117 +0x2ff
  github.com/cockroachdb/cockroach/pkg/server_test.TestMain()
      /go/src/github.com/cockroachdb/cockroach/pkg/server/main_test.go:28 +0x156
  main.main()
      _testmain.go:288 +0x223

Goroutine 403 (finished) created at:
  github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).initConnEx()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:174 +0x595
  github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).execInternal()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:446 +0x7ee
  github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).queryInternal()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:241 +0x14f
  github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).QueryEx()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:216 +0x14f
  github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).QueryRowEx()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:271 +0x14f
  github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).QueryRow()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:255 +0x2b6
  github.com/cockroachdb/cockroach/pkg/sql/sqlliveness/slstorage.(*Storage).Update.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/sqlliveness/slstorage/slstorage.go:286 +0x2b7
  github.com/cockroachdb/cockroach/pkg/kv.(*DB).Txn.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:707 +0x5e
  github.com/cockroachdb/cockroach/pkg/kv.(*Txn).exec()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/txn.go:811 +0xe9
  github.com/cockroachdb/cockroach/pkg/kv.(*DB).Txn()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:706 +0x11c
  github.com/cockroachdb/cockroach/pkg/sql/sqlliveness/slstorage.(*Storage).Update()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/sqlliveness/slstorage/slstorage.go:285 +0x126
  github.com/cockroachdb/cockroach/pkg/sql/sqlliveness/slinstance.(*Instance).extendSession()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/sqlliveness/slinstance/slinstance.go:161 +0x3b0
  github.com/cockroachdb/cockroach/pkg/sql/sqlliveness/slinstance.(*Instance).heartbeatLoop()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/sqlliveness/slinstance/slinstance.go:207 +0x2e4
  github.com/cockroachdb/cockroach/pkg/sql/sqlliveness/slinstance.(*Instance).heartbeatLoop-fm()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/sqlliveness/slinstance/slinstance.go:183 +0x55
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:222 +0x1b8
==================
FAIL	github.com/cockroachdb/cockroach/pkg/server	312.828s

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestTelemetrySQLStatsIndependence PKG=./pkg/server TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestTelemetrySQLStatsIndependence.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(server).TestTelemetrySQLStatsIndependence failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2195924&tab=buildLog) on [master@ebd5c732f83009acc9c6f5859ca95e74e5453a1c](https://github.com/cockroachdb/cockroach/commits/ebd5c732f83009acc9c6f5859ca95e74e5453a1c):

```
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:960 +0x651
  testing.runTests.func1()
      /usr/local/go/src/testing/testing.go:1202 +0xa6
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:909 +0x199
  testing.runTests()
      /usr/local/go/src/testing/testing.go:1200 +0x521
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:1117 +0x2ff
  github.com/cockroachdb/cockroach/pkg/server_test.TestMain()
      /go/src/github.com/cockroachdb/cockroach/pkg/server/main_test.go:28 +0x156
  main.main()
      _testmain.go:288 +0x223

Goroutine 969 (finished) created at:
  github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).initConnEx()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:174 +0x595
  github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).execInternal()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:446 +0x7f4
  github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).queryInternal()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:241 +0x14f
  github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).QueryEx()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:216 +0x14f
  github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).QueryRowEx()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:271 +0x14f
  github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).QueryRow()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:255 +0x2b6
  github.com/cockroachdb/cockroach/pkg/sql/sqlliveness/slstorage.(*Storage).deleteExpiredSessions.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/sqlliveness/slstorage/slstorage.go:232 +0x216
  github.com/cockroachdb/cockroach/pkg/kv.(*DB).Txn.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:707 +0x5e
  github.com/cockroachdb/cockroach/pkg/kv.(*Txn).exec()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/txn.go:811 +0xe9
  github.com/cockroachdb/cockroach/pkg/kv.(*DB).Txn()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:706 +0x11c
  github.com/cockroachdb/cockroach/pkg/sql/sqlliveness/slstorage.(*Storage).deleteExpiredSessions()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/sqlliveness/slstorage/slstorage.go:231 +0xfc
  github.com/cockroachdb/cockroach/pkg/sql/sqlliveness/slstorage.(*Storage).deleteSessionsLoop()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/sqlliveness/slstorage/slstorage.go:224 +0x225
  github.com/cockroachdb/cockroach/pkg/sql/sqlliveness/slstorage.(*Storage).deleteSessionsLoop-fm()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/sqlliveness/slstorage/slstorage.go:208 +0x55
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:347 +0x162
==================
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestTelemetrySQLStatsIndependence985282058
--- FAIL: TestTelemetrySQLStatsIndependence (5.69s)
    test_log_scope.go:154: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestTelemetrySQLStatsIndependence985282058
    test_log_scope.go:63: use -show-logs to present logs inline
    testing.go:853: race detected during execution of test

```

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-race -parallel=2

```
make stressrace TESTS=TestTelemetrySQLStatsIndependence PKG=./pkg/server TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestTelemetrySQLStatsIndependence.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/sql/app_stats.go
+++ b/pkg/sql/app_stats.go
@@ -495,18 +495,21 @@ func (s *sqlStats) getStmtStats(
 				}
 			}
 			if ok {
+				stats.Lock()
+				data := stats.data
+				distSQLUsed := stats.distSQLUsed
+				vectorized := stats.vectorized
+				stats.Unlock()
+
 				k := roachpb.StatementStatisticsKey{
 					Query:       maybeScrubbed,
-					DistSQL:     stats.distSQLUsed,
+					DistSQL:     distSQLUsed,
 					Opt:         true,
-					Vec:         stats.vectorized,
+					Vec:         vectorized,
 					ImplicitTxn: q.implicitTxn,
 					Failed:      q.failed,
 					App:         maybeHashedAppName,
 				}
-				stats.Lock()
-				data := stats.data
-				stats.Unlock()
 
 				if scrub {
 					// Quantize the counts to avoid leaking information that way.

